File: osdev_18/src/driver/Keyboard.zig
const std = @import("std");
const Console = @import("Console.zig");
const isr = @import("../isr.zig");
const utils = @import("../utils.zig");

// Keyboard state
var lshift = false;

fn handler(_: isr.Registers) void {
    const scancode = utils.inb(0x60);
    switch (scancode) {
        0x0B => if (lshift) Console.write(")") else Console.write("0"),
        0x02 => if (lshift) Console.write("!") else Console.write("1"),
        0x03 => if (lshift) Console.write("@") else Console.write("2"),
        0x04 => if (lshift) Console.write("#") else Console.write("3"),
        0x05 => if (lshift) Console.write("$") else Console.write("4"),
        0x06 => if (lshift) Console.write("%") else Console.write("5"),
        0x07 => if (lshift) Console.write("^") else Console.write("6"),
        0x08 => if (lshift) Console.write("&") else Console.write("7"),
        0x09 => if (lshift) Console.write("*") else Console.write("8"),
        0x0A => if (lshift) Console.write("(") else Console.write("9"),
        0x0C => if (lshift) Console.write("_") else Console.write("-"),
        0x35 => if (lshift) Console.write("?") else Console.write("/"),
        0x2B => if (lshift) Console.write("\\") else Console.write("|"),
        0x0D => if (lshift) Console.write("=") else Console.write("+"),
        0x1A => if (lshift) Console.write("{") else Console.write("["),
        0x1B => if (lshift) Console.write("}") else Console.write("]"),
        0x10 => if (lshift) Console.write("Q") else Console.write("q"),
        0x11 => if (lshift) Console.write("W") else Console.write("w"),
        0x12 => if (lshift) Console.write("E") else Console.write("e"),
        0x13 => if (lshift) Console.write("R") else Console.write("r"),
        0x14 => if (lshift) Console.write("T") else Console.write("t"),
        0x15 => if (lshift) Console.write("Y") else Console.write("y"),
        0x16 => if (lshift) Console.write("U") else Console.write("u"),
        0x17 => if (lshift) Console.write("I") else Console.write("i"),
        0x18 => if (lshift) Console.write("O") else Console.write("o"),
        0x19 => if (lshift) Console.write("P") else Console.write("p"),
        0x1E => if (lshift) Console.write("A") else Console.write("a"),
        0x1F => if (lshift) Console.write("S") else Console.write("s"),
        0x20 => if (lshift) Console.write("D") else Console.write("d"),
        0x21 => if (lshift) Console.write("F") else Console.write("f"),
        0x22 => if (lshift) Console.write("G") else Console.write("g"),
        0x23 => if (lshift) Console.write("H") else Console.write("h"),
        0x24 => if (lshift) Console.write("J") else Console.write("j"),
        0x25 => if (lshift) Console.write("K") else Console.write("k"),
        0x26 => if (lshift) Console.write("L") else Console.write("l"),
        0x2C => if (lshift) Console.write("Z") else Console.write("z"),
        0x2D => if (lshift) Console.write("X") else Console.write("x"),
        0x2E => if (lshift) Console.write("C") else Console.write("c"),
        0x2F => if (lshift) Console.write("V") else Console.write("v"),
        0x30 => if (lshift) Console.write("B") else Console.write("b"),
        0x31 => if (lshift) Console.write("N") else Console.write("n"),
        0x32 => if (lshift) Console.write("M") else Console.write("m"),
        0x39 => Console.write(" "),
        0x1C => Console.write("\n"),
        0x2A => lshift = true,
        0xAA => lshift = false,
        else => {
            // Print as hex
            // const index = [16]u8{ '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F' };
            // Console.write(&.{ ' ', '0', 'x', index[(scancode & 0xF0) >> 4], index[(scancode & 0x0F)], ' ' });
        },
    }
}

pub fn init() void {
    isr.setHandler(isr.IRQ1, handler);
    utils.outb(0x21, 0xFD);
    utils.outb(0xA1, 0xFF);
}

File: group_23/src/kernel/src/keyboard.c
#include "../include/keyboard.h"
#include "../include/isr.h"
#include "../include/monitor.h"
#include "../include/common.h"
#include <stdint.h>
#include <string.h>

#define SC_MAX 57
#define BACKSPACE 0x0E
#define ENTER 0x1C
static char key_buffer[256];

// code from cfenollosa tutorial: https://github.com/cfenollosa/os-tutorial/blob/master/23-fixes/drivers/keyboard.c



static void keyboard_callback(registers_t *regs){
    uint8_t scancode= inb(0x60);
    if(scancode>SC_MAX) return;
    if(scancode==BACKSPACE){
        monitor_put(key_buffer);
        
    }
    
    else{

        monitor_put('\n');
        monitor_write("Keyboard scancode: ");
        
        print_letter(scancode);
        
    }
    
}

void init_keyboard() {
    

    register_interrupt_handler(IRQ1, keyboard_callback);
    
}



void print_letter(uint8_t scancode) 
{
    switch (scancode) 
    {
        case 0x0:
            monitor_write("ERROR");
            break;
        case 0x1:
            monitor_write("ESC");
            break;
        case 0x2:
            monitor_put('1');
            break;
        case 0x3:
            monitor_put('2');
            break;
        case 0x4:
            monitor_put('3');
            break;
        case 0x5:
            monitor_put('4');
            break;
        case 0x6:
            monitor_put('5');
            break;
        case 0x7:
            monitor_put('6');
            break;
        case 0x8:
            monitor_put('7');
            break;
        case 0x9:
            monitor_put('8');
            break;
        case 0x0A:
            monitor_put('9');
            break;
        case 0x0B:
            monitor_put('0');
            break;
        case 0x0C:
            monitor_put('-');
            break;
        case 0x0D:
            monitor_put('+');
            break;
        case 0x0E:
            monitor_write("Backspace");
            break;
        case 0x0F:
            monitor_write("Tab");
            break;
        case 0x10:
            monitor_put('Q');
            break;
        case 0x11:
            monitor_put('W');
            break;
        case 0x12:
            monitor_put('E');
            break;
        case 0x13:
            monitor_put('R');
            break;
        case 0x14:
            monitor_put('T');
            break;
        case 0x15:
            monitor_put('Y');
            break;
        case 0x16:
            monitor_put('U');
            break;
        case 0x17:
            monitor_put('I');
            break;
        case 0x18:
            monitor_put('O');
            break;
        case 0x19:
            monitor_put('P');
            break;
        case 0x1A:
            monitor_put('[');
            break;
        case 0x1B:
            monitor_put(']');
            break;
        case 0x1C:
            monitor_write("ENTER");
            break;
        case 0x1D:
            monitor_write("LCtrl");
            break;
        case 0x1E:
            monitor_put('A');
            break;
        case 0x1F:
            monitor_put('S');
            break;
        case 0x20:
            monitor_put('D');
            break;
        case 0x21:
            monitor_put('F');
            break;
        case 0x22:
            monitor_put('G');
            break;
        case 0x23:
            monitor_put('H');
            break;
        case 0x24:
            monitor_put('J');
            break;
        case 0x25:
            monitor_put('K');
            break;
        case 0x26:
            monitor_put('L');
            break;
        case 0x27:
            monitor_put(';');
            break;
        case 0x28:
            monitor_put('\'');
            break;
        case 0x29:
            monitor_put('`');
            break;
        case 0x2A:
            monitor_write("LShift");
            break;
        case 0x2B:
            monitor_put('\\');
            break;
        case 0x39:
            monitor_write("Spacebar");
            break;
        default:
            /* 'keyboard.c' defines all scancodes above as 'ERROR' */
            if (scancode <= 0x7f){
                monitor_write("Unknown key down");
            }
            else if (scancode <= 0x39 + 0x80){
                monitor_write("key up");
                print_letter(scancode - 0x80);
            }
            else{
                monitor_write("Unknown key up");
            }
            break;
        
    }
}
